### ISSUE TITLE
Core: Ban Yoda conditions, loose comparisons and assignments in conditions

### ISSUE THREAD

**Issue by jrfnl** [WordPress/WordPress-Coding-Standards#1624]
## Is your feature request related to a problem?

_Yoda conditions_ are the practice to reverse how comparisons in conditions are written, having the value being compared to on the left hand of the condition and the variable on the right hand, to prevent accidentally assigning something to a variable in a condition.
```php
// This condition:
if ( $variable === 'value' ) {}

// ...would be written like this in Yoda style:
if ( 'value' === $variable ) {}

// ... to prevent this:
if ( $variable = 'value' ) {}
```

**The problem with this is that code readability is decreased significantly to prevent a beginner's mistake which shouldn't be made in the first place and the chances of which are small when using strict comparisons anyhow.**

Now, the WordPress coding standards contains a few sniffs related to this:
* `Core`: `WordPress.PHP.YodaConditions` (`error`) - demands the use of Yoda conditions as per [the handbook](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#yoda-conditions).
    However, this sniff only looks at _comparison operators_, i.e. `==`, `===`, `!=` and `!==`, **not** at _assignment operators_, like `=`, so in effect, **it doesn't actually prevent assignments in conditions in the first place**.
* `Extra`: `WordPress.CodeAnalysis.AssignmentInCondition` (`warning`) - warns against assignments found in conditions of control structures as well as, in as far possible, assignments in the condition part of ternaries.
* `Extra`: `WordPress.PHP.StrictComparisons` (`warning`) - warns against using loose comparisons.
* `Extra`: `WordPress.PHP.StrictInArray` (`warning`) - warns against using loose comparisons in array comparison functions.

Note: none of these sniffs currently contain (auto-)fixers.

## Describe the solution you'd like

At this moment, I'd like to propose the following:
* Drop the handbook rule to demand Yoda conditions.
* Remove the `WordPress.PHP.YodaConditions` sniff - both from the `Core` ruleset and from the repo (initially deprecated and emptied out, to be removed in the next major release).
* Move the `WordPress.CodeAnalysis.AssignmentInCondition` sniff to the `Core` ruleset and change it from a `warning` to an `error`.
* Move the `WordPress.PHP.StrictComparisons` and `WordPress.PHP.StrictInArray` sniffs to the `Core` ruleset, but leave these as a `warning`.

### Impact
* At this point, WPCS would have no opinion on whether or not to use Yoda conditions.
* I expect that the `WordPress.CodeAnalysis.AssignmentInCondition` sniff would introduce a limited set of newly detected issues for the WP Core code.
* I expect that the introduction of the `WordPress.PHP.StrictComparisons` and `WordPress.PHP.StrictInArray` sniffs would introduce a large number of new warnings for the WP Core code base.

Once these changes have been made to WPCS and Core has upgraded, contributors can be encouraged to start looking at the newly detected issues and to fix up the existing (non-strict) comparisons.
This should be done carefully as changing the type of comparison is a behavioural change.
It is recommended to make sure that all changes are covered by unit tests before making this type of change.

## Next phase

As a next step, I'd like to propose the handbook to be adjusted to _forbid_ Yoda conditions.

A new sniff will need to be written to detect the use of Yoda Conditions, throw an error and where possible, auto-fix this.

Alternatively it could be considered to add the `Slevomat Coding Standard` as a dependency to WPCS and use the existing sniff in that standard.
The downsides of that would be:
- A new external dependency making installation of WPCS with anything but Composer more fiddly.
- No control over the sniff, making getting bugfixes and such accepted slightly more fiddly.

## Additional context (optional)

Some discussion about this has been had in other channels. I'm quoting some relevant parts of that discussion here.

> @jrfnl Regarding Yoda conditions - we've talked about this before. IMO this rule should be dropped in favour of the "No assignment in condition" rule and the associated sniff `WordPress/Generic.CodeAnalysis.AssignmentInCondition` which is currently in `Extra` should be moved to `Core`.
>
> At the time you asked me to investigate whether there is a sniff available out there which could revert existing Yoda conditions. The answer to that is "Yes". The external [Slevomat Coding Standard](https://github.com/slevomat/coding-standard#slevomatcodingstandardcontrolstructuresdisallowyodacomparisonrequireyodacomparison-) contains a sniff which can auto-fix Yoda to non-Yoda. Tests could (should) be run to see how well it works, but this sniff could be used in a one-time action to move away from Yoda. Note: if the sniff would not be good enough/would turn out to be buggy, I'm happy to contribute fixes to the Slevomat standard just to make moving away from Yoda possible.
>
> Having said that, as the tooling is available which would allow us to move away from Yoda and still prevent assignments in conditions, I would strongly suggest now is a good time to take a decision on this. 

Source: https://core.trac.wordpress.org/ticket/45934#comment:14

>  @pento Yah, I agree that `WordPress/Generic.CodeAnalysis.AssignmentInCondition` should be moved into `Core`, violations should be fixed, and then Yoda conditions should be removed. There's some discussion on [#42885](https://core.trac.wordpress.org/ticket/42885) about Yoda conditions, too.
>
> Notably, the [JS team are planning on removing guidance on Yoda conditions](https://make.wordpress.org/core/2018/12/18/proposed-revisions-to-javascript-coding-standards/), I'm inclined to disallow Yoda conditions, rather than simply allow either style, though. 

Source: https://core.trac.wordpress.org/ticket/45934#comment:15

Opinions ?

**Comment by pento** [comment]
I'm in favour of this change.

There are 559 violations of `WordPress.CodeAnalysis.AssignmentInCondition` in WordPress Core.

This seems like something that could be auto-fixed, though. Move the assignment to before the offending block, and replace the assignment in the condition to just the variable name.

For example:

```php
if ( $foo = bar() ) {
	// ...
} elseif ( $baz = bat() ) {
	// ...
}
```

Would become:

```php
$foo = bar();
$baz = bat();
if ( $foo ) {
	// ...
} elseif ( $baz ) {
	// ...
}
```

This could be a problem if `bat()` should only over be run if `bar()` returned something false-y, due to side-effects or whatever. Arguably such code is fundamentally broken, anyway. 🙂

It would also break on this code:

```php
if ( $foo = bar() ) {
	// ...
} elseif ( $foo = bat() ) {
	// ...
}
```

**Comment by jrfnl** [comment]
@pento IMO assignments in condition should not be auto-fixed.
A tool can not determine whether this is an assignment which should be _moved out of the condition_ or an _accidental assignment_ while a comparison was intended.

While I hope and presume that that last type doesn't exist in Core, we won't know for sure until we look at them.

Also, the complexity would be too high for an auto-fixer.
Think:
```php
if ( isset($foo) && ! ($bar = function_call($foo)) && $bar > 3) {}
```
Or:
```php
while ( ($foo = function_call($bar)) !== false) {
    // Do something
}
```

Note: for assignments in `while` conditions, the sniff has a separate error code to allow selectively whitelisting those, as at times, that is a valid way of writing the code, though changing the code to a `do {} while()` or other type of control structure should probably be preferred.

**Comment by 2ndkauboy** [comment]
I agree that removing the Yoda Condition check could be removed from the Core ruleset, but it should not be removed from the "PHP Coding Standards" in the handbook and should still be checked in the WordPress and/or WordPress-Extra ruleset.

**Comment by NielsdeBlaauw** [comment]
> A new sniff will need to be written to detect the use of Yoda Conditions, throw an error and where possible, auto-fix this.
> 
> Alternatively it could be considered to add the `Slevomat Coding Standard` as a dependency to WPCS and use the existing sniff in that standard.

Would it be an option to add detection sniff in WPCS, and use a [composer suggest](https://getcomposer.org/doc/04-schema.md#suggest) with message that `SlevomatCodingStandard.ControlStructures.DisallowYodaComparison can be used as a fixer`. 

Looking at the Slevomat Coding standard it appears the fixer is far more complex than the sniff part of the implementation. 

I think most projects would likely only be using it once during onboarding/initial scan. After that it's enough to be warned about the instance, an it would be natural to not type yoda conditions in the first place.

**Comment by JJJ** [comment]
Anecdotally, I believe Yoda conditions improve code quality & readability.

They are a decades old practice with a proven track record at avoiding accidental assignment errors.

Removing them is removing the steel toes from our workboots because they aren’t comfortable.

Evidently my feelings here aren’t the currently popular ones, but I’m a big -1 on switching away from them.

**Comment by GaryJones** [comment]
> Removing them is removing the steel toes from our workboots because they aren’t comfortable.

The presence of the assignment in condition check, means we've already removed the chance of anything heavy landing on our feet, so the steel toe caps would be redundant 😄

**Comment by JPry** [comment]
I'm with @JJJ on this. I don't see any need or improvement to switching away from Yoda conditions. 

> The problem with this is that code readability is decreased significantly

How is readability decreased, especially "significantly"? Are there references somewhere that aren't included here that can be linked to?

Even if the preference against Yoda conditions continues, I think there's no reason to completely reverse stance and start saying that Yoda conditions are bad practice.

**Comment by pento** [comment]
Having thought about this some more, as well as discussing it with a range of different people, I'm not sure that banning Yoda conditions entirely is the most reasonable next step.

@JJJ was kind enough to go into the detail of how he finds Yoda conditions easier than non-Yoda to understand. @nacin mentioned the special case of checking `false === strpos( ... )`, and the like.

So, I'd like to explore this further before making a change. @davidakennedy recently published [an interesting post about designing for accessibility](https://automattic.design/2019/03/04/better-conversations-about-accessibility/), which I think also applies to this conversation. Let's work out what our constraints are, then make a change that works as well as possible for everyone.

I'm writing a post for make/core about WordPress coding standards changes with relation to the PHP version bump, so I'll include a call for feedback on the topic of Yoda conditions there.

**Comment by jrfnl** [comment]
@pento What about the initial step ?
* Removing the yoda conditions rule, _without replacing it with a ban_ (i.e. opinion, both yoda and non-yoda allowed);
* Banning assignments in conditions;
* Recommending strict comparisons.

Should any or all of these be actioned yet ?

Note: each of these three points can be actioned independently of the others.

**Comment by GaryJones** [comment]
> * Removing the yoda conditions rule, _without replacing it with a ban_ (i.e. opinion, both yoda and non-yoda allowed);

To state what might be obvious - the decision to follow Yoda conditions in _Core_ could be set in Core's ruleset or `WordPress-Core`, even if `WordPress-Extra` removes this requirement.

**Comment by pento** [comment]
@jrfnl: Still working on the post about these. 🙂 Let's not action any of them yet, I suspect that 2 and 3 can happen quickly, but 1 will need some more information on the best way forward.

**Comment by jrfnl** [comment]
@pento OK, thanks for the heads-up. I will remove this from the WPCS 2.1.0 milestone for now.

**Comment by CurtisBelt** [comment]
@pento

> I'm writing a post for make/core about WordPress coding standards changes with relation to the PHP version bump, so I'll include a call for feedback on the topic of Yoda conditions there.

Can you provide link to the post you mentioned? Just trying to trace the progress of this issue. Thanks!

**Comment by pento** [comment]
@CurtisBelt: The post I was referring to was published [here](https://make.wordpress.org/core/2019/03/26/coding-standards-updates-for-php-5-6/), with a [followup posted a little later](https://make.wordpress.org/core/2019/07/12/php-coding-standards-changes/).

You'll note that Yoda conditions weren't mentioned in either post. On further reflection while writing the posts, that topic had the potential to be too much of a distraction from the more immediate changes that could happen. Unfortunately, I never got back to writing a post just for the Yoda conditions discussion.

I'm currently taking a break from working on WordPress Core, so I don't expect to have time to look at this in the near future. Given that the next step is to explore the topic further in a wider discussion, this is something that anyone interested could drive, it certainly doesn't require my input to move it along. I would suggest bringing it up at the next [core dev chat](https://make.wordpress.org/core/tag/devchat/) as a way to gauge interest in a wider discussion.

**Comment by ggedde** [comment]
I really just think the Yoda conditions need to be removed. 
I am all for keeping loose comparisons and assignments in conditions warnings.
I agree with @GaryJones that by having loose comparisons and assignments in conditions is enough to also prevent accidental typos with "=" and "==".

**Comment by Luc45** [comment]
Yoda conditions can be automatically converted to regular conditions with [PHP CS Fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer):

```
'yoda_style' => [
    'equal' => false,
    'identical' => false,
    'less_and_greater' => false,
],
```

**Comment by mitogh** [comment]
Have we considered the fact to only enforce it when using 2 equal operators and ignore it when using a strict comparison with 3 equal operators instead?

```
// Enforce it if
if ( $value == false ) {
}

// Ignore if
if ( $value === false ) {
}
```

Primarily because based on:

> In the above example, if you omit an equals sign (admit it, it happens even to the most seasoned of us), you’ll get a parse error, because you can’t assign to a constant like true. If the statement were the other way around ( $the_force = true ), the assignment would be perfectly valid, returning 1, causing the if statement to evaluate to true, and you could be chasing that bug for a while.


[Source](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#yoda-conditions) 

When using strict operator the premise from the docs is no longer valid as you are ensuring that won't happen :)

**Comment by sirbrillig** [comment]
> Have we considered the fact to only enforce it when using 2 equal operators and ignore it when using a strict comparison with 3 equal operators instead?

That's a reasonable idea, but in that case why not enforce it only if using _one_ equals operator (which really is the bug this rule is trying to prevent)?

```
// Enforce it if
if ( $value = false ) {
}

// Ignore it if
if ( $value == false ) {
}

// Ignore if
if ( $value === false ) {
}
```

The standard is there to make sure that a typo isn't made on a comparison, but automated linting can do that instead, and generally better than humans.

**Comment by mitogh** [comment]
> That's a reasonable idea, but in that case, why not enforce it only if using one equals operator (which really is the bug this rule is trying to prevent)?

Mostly for backward compatibility and to make sure when updating to a new rule system existing rules are no longer valid or considered instead. But I do agree if this can be automated I don't see a practical reason why not to replicate it in 2 comparison operators as well. 

> The standard is there to make sure that a typo isn't made on a comparison, but automated linting can do that instead, and generally better than humans.

:100: on board with this idea.

**Comment by jrfnl** [comment]
May I suggest reading the original proposal ? As that's kind of what was originally suggested.

**Comment by sirbrillig** [comment]
There must be some axiom that says that if a discussion thread gets long enough it will loop back over itself. 😁

**Comment by mitogh** [comment]
> May I suggest reading the original proposal ? As that's kind of what was originally suggested.

Well partially, from the main proposal is actually requesting a ban which was updated here: https://github.com/WordPress/WordPress-Coding-Standards/issues/1624#issuecomment-470821000

Maybe should the main ticket description should be updated for a more simple approach instead as you described here: https://github.com/WordPress/WordPress-Coding-Standards/issues/1624#issuecomment-470821000

**Comment by Luc45** [comment]
+1 non-yoda conditions. I find yoda counter-intuitive, as imagine you could.

Perhaps we could open a poll to settle this?

**Comment by JJJ** [comment]
WordPress standardized on Yoda conditions 15 _years_ ago.

Every plugin & theme has inherited that standard, resulting in verifiably millions of lines of PHP code that are written this way.

A timeline of future events after this change happens:

* WordPress changes everything internally
* Several plugins & themes immediately adopt
* Some plugins & themes slowly adopt
* Most plugins & themes never adopt

If this changes, the larger WordPress landscape is left littered with the simplest code imaginable (variable comparisons) looking _different_ everywhere, in _every_ project, all over the web. A few people will be happy and feel accomplished, and everyone will suffer.

Changing this results in code-slurry across the web that is marginally easier for a vocal minority to ingest, but will be highly disorienting for everyone else – I.E. people like my brilliant wife who is, as I write this, learning PHP for her very first time and is constantly tripping over (and frustrated by) uselessly differing code _structure_.

It will be a mistake to the entire ecosystem to intentionally **invert** such a prolific WordPress'ism.

**Comment by anomiex** [comment]
> If this changes, the larger WordPress landscape is left littered with the simplest code imaginable (variable comparisons) looking _different_ everywhere, in every project, all over the web.

On the other hand, the larger PHP ecosystem seems to generally be non-Yoda. From this perspective WordPress seems like the outlier.

Then again, the larger PHP ecosystem has also standardized on PSR-4 for naming files containing classes; camelCase for method and variable names; and short array syntax. I've not seen any other project that requires aligning the `=>` in arrays or `=` in assignments. And few projects seem to have clung to PHP 5.6 for so long.

**Comment by dingo-d** [comment]
@JJJ just because something was added 15 years ago, doesn't mean it should stay this way forever.

I don't think that the argument that various plugins/themes/projects out there having different coding standards (the reality is that the majority doesn't have any) should prevent changes in the WordPress core coding standards.

If anything it would be better if we are doing what the rest of the PHP community is doing and see if we can somehow align with their standard.

That way WordPress wouldn't be _the odd one out_ IMO.

Plus, the point of using automated tools such as PHPCS is to prevent issues that Yoda conditions solved back when we didn't have such tools (or didn't use them)
.

**Comment by Luc45** [comment]
> A few people will be happy and feel accomplished, and everyone will suffer.

> easier for a vocal minority to ingest, but will be highly disorienting for everyone else

I have downloaded almost all the 93 965 plugins in the WordPress Plugin Repository (Due to network gaps, I was able to download 82 450 plugins directly from wordpress.org servers (93GB)) and ran PHPCS to find the usage of Yoda vs Non Yoda conditions. This provides us with data that brings a pragmatic insight that can be beneficial to the discussion.

**I will try to be as transparent, pragmatic, and scientific as possible, to describe every inch of my methodology so that it can be peer-reviewed to make sure it does not contain any bias.**

This is the phpcs.xml that I have used:

```xml
<?xml version="1.0"?>
<ruleset name="Yoda Analyzer">
    <config name="installed_paths" value="/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/phpcs/vendor/slevomat/coding-standard"/>

    <file>/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/plugins/0-delay-late-caching-for-feeds</file>
    <file>/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/plugins/0-errors</file>
    <file>/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/plugins/001-prime-strategy-translate-accelerator</file>
    <file>/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/plugins/002-ps-custom-post-type</file>
    <file>/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/plugins/011-ps-custom-taxonomy</file>
    <file>/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/plugins/012-ps-multi-languages</file>
    <file>/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/plugins/028-ps-combine-taxonomy-children</file>
    <file>/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/plugins/030-ps-display-upload-path-for-wp35</file>
    <file>/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/plugins/03talk-community-conference</file>
    <file>/home/lucas/local/data/www/plugin-directory/WordPress-Plugin-Directory-Slurper/plugins/0gravatar</file>
    <!-- ... etc, until all plugins are scanned (in batches) -->

    <exclude-pattern>*/vendor/*</exclude-pattern>
    <exclude-pattern>*/node_modules/*</exclude-pattern>
    <exclude-pattern>*/freemius/*</exclude-pattern>
    <exclude-pattern>*/vendor_prefixed/*</exclude-pattern>
    <exclude-pattern>*/composer/*</exclude-pattern>

    <arg name="extensions" value="php"/>

    <arg name="parallel" value="12"/>

    <rule ref="SlevomatCodingStandard.ControlStructures.DisallowYodaComparison"/>
    <rule ref="SlevomatCodingStandard.ControlStructures.RequireYodaComparison"/>
</ruleset>
```

I have tried to do my best to exclude any vendor code, such as packages that might be installed using composer, and focus only on the WordPress PHP code of the plugin.

- [This is](https://github.com/slevomat/coding-standard/blob/master/tests/Sniffs/ControlStructures/data/disallowYodaComparisonErrors.php) what the sniff `SlevomatCodingStandard.ControlStructures.DisallowYodaComparison` considers Yoda code.
- [This is](https://github.com/slevomat/coding-standard/blob/master/tests/Sniffs/ControlStructures/data/requireYodaComparisonErrors.php) what the sniff `SlevomatCodingStandard.ControlStructures.RequireYodaComparison` considers non-Yoda code.

I've used PHPCS's `--report=source` option, which gives me a result such as this (this example is for the first 10k scans):

```
PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
--------------------------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                                COUNT
--------------------------------------------------------------------------------------------
[x] Slevomat  Control structures  Require yoda comparison required yoda comparison     426417
[x] Slevomat  Control structures  Disallow yoda comparison disallowed yoda comparison  100539
[ ] Internal                      No code found                                        1851
[ ] Internal                      Line endings                                         1475
[ ] Internal                      Exception                                            6
--------------------------------------------------------------------------------------------
A TOTAL OF 530288 SNIFF VIOLATIONS WERE FOUND IN 5 SOURCES
--------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SOURCES AUTOMATICALLY (526479 VIOLATIONS IN TOTAL)
--------------------------------------------------------------------------------------------

Time: 7 mins, 3.01 secs; Memory: 56.24MB
```

`Require yoda comparison required yoda comparison` is the rule where we require Yoda conditions, and the count is the number of non-yoda conditions if found. `Disallow yoda comparison disallowed yoda comparison` is the rule where we disallow Yoda conditions, and the count is the yoda conditions it found.

I had to split the PHPCS run into batches, as it was too much for my PC to process in one go. To do that, I wrote a small script that generates a phpcs.xml with plugins within a range.

<details>
  <summary>Results per batch</summary>
  
**Plugins 0 to 10k:**
```
➜  phpcs git:(master) ✗ cat 0_10000.txt 

PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
--------------------------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                                COUNT
--------------------------------------------------------------------------------------------
[x] Slevomat  Control structures  Require yoda comparison required yoda comparison     426417
[x] Slevomat  Control structures  Disallow yoda comparison disallowed yoda comparison  100539
[ ] Internal                      No code found                                        1851
[ ] Internal                      Line endings                                         1475
[ ] Internal                      Exception                                            6
--------------------------------------------------------------------------------------------
A TOTAL OF 530288 SNIFF VIOLATIONS WERE FOUND IN 5 SOURCES
--------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SOURCES AUTOMATICALLY (526479 VIOLATIONS IN TOTAL)
--------------------------------------------------------------------------------------------

Time: 7 mins, 3.01 secs; Memory: 56.24MB
```

**Plugins 10k to 20k:**
```
➜  phpcs git:(master) ✗ cat 10000_20000.txt 

PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
--------------------------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                                COUNT
--------------------------------------------------------------------------------------------
[x] Slevomat  Control structures  Require yoda comparison required yoda comparison     418345
[x] Slevomat  Control structures  Disallow yoda comparison disallowed yoda comparison  91076
[ ] Internal                      No code found                                        2743
[ ] Internal                      Line endings                                         1233
--------------------------------------------------------------------------------------------
A TOTAL OF 513397 SNIFF VIOLATIONS WERE FOUND IN 4 SOURCES
--------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SOURCES AUTOMATICALLY (509083 VIOLATIONS IN TOTAL)
--------------------------------------------------------------------------------------------

Time: 7 mins, 0.65 secs; Memory: 54.12MB
```

**Plugins 20k to 30k:**
```
➜  phpcs git:(master) ✗ cat 20000_30000.txt 

PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
--------------------------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                                COUNT
--------------------------------------------------------------------------------------------
[x] Slevomat  Control structures  Require yoda comparison required yoda comparison     445222
[x] Slevomat  Control structures  Disallow yoda comparison disallowed yoda comparison  92608
[ ] Internal                      Line endings                                         2012
[ ] Internal                      No code found                                        1590
[ ] Internal                      Exception                                            2
--------------------------------------------------------------------------------------------
A TOTAL OF 541434 SNIFF VIOLATIONS WERE FOUND IN 5 SOURCES
--------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SOURCES AUTOMATICALLY (537569 VIOLATIONS IN TOTAL)
--------------------------------------------------------------------------------------------

Time: 7 mins, 37.34 secs; Memory: 54.27MB
```

**Plugins 30k to 40k:**
```
➜  phpcs git:(master) ✗ cat 30000_40000.txt 

PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
--------------------------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                                COUNT
--------------------------------------------------------------------------------------------
[x] Slevomat  Control structures  Require yoda comparison required yoda comparison     458963
[x] Slevomat  Control structures  Disallow yoda comparison disallowed yoda comparison  101565
[ ] Internal                      Line endings                                         3784
[ ] Internal                      No code found                                        2689
[ ] Internal                      Exception                                            7
--------------------------------------------------------------------------------------------
A TOTAL OF 567008 SNIFF VIOLATIONS WERE FOUND IN 5 SOURCES
--------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SOURCES AUTOMATICALLY (560188 VIOLATIONS IN TOTAL)
--------------------------------------------------------------------------------------------

Time: 6 mins, 50.86 secs; Memory: 58.91MB
```

**Plugins 40k to 50k:**
```
➜  phpcs git:(master) ✗ cat 40000_50000.txt 

PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
--------------------------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                                COUNT
--------------------------------------------------------------------------------------------
[x] Slevomat  Control structures  Require yoda comparison required yoda comparison     479307
[x] Slevomat  Control structures  Disallow yoda comparison disallowed yoda comparison  96621
[ ] Internal                      Line endings                                         3196
[ ] Internal                      No code found                                        1631
[ ] Internal                      Exception                                            1
--------------------------------------------------------------------------------------------
A TOTAL OF 580756 SNIFF VIOLATIONS WERE FOUND IN 5 SOURCES
--------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SOURCES AUTOMATICALLY (575509 VIOLATIONS IN TOTAL)
--------------------------------------------------------------------------------------------

Time: 7 mins, 33.55 secs; Memory: 62.96MB
```

**Plugins 50k to 60k:**
```
➜  phpcs git:(master) ✗ cat 50000_60000.txt 

PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
--------------------------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                                COUNT
--------------------------------------------------------------------------------------------
[x] Slevomat  Control structures  Require yoda comparison required yoda comparison     364109
[x] Slevomat  Control structures  Disallow yoda comparison disallowed yoda comparison  82842
[ ] Internal                      No code found                                        1617
[ ] Internal                      Line endings                                         1123
--------------------------------------------------------------------------------------------
A TOTAL OF 449691 SNIFF VIOLATIONS WERE FOUND IN 4 SOURCES
--------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SOURCES AUTOMATICALLY (446646 VIOLATIONS IN TOTAL)
--------------------------------------------------------------------------------------------

Time: 6 mins, 25.88 secs; Memory: 47.55MB

```

**Plugins 60k to 70k:**
```
➜  phpcs git:(master) ✗ cat 60000_70000.txt

PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
--------------------------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                                COUNT
--------------------------------------------------------------------------------------------
[x] Slevomat  Control structures  Require yoda comparison required yoda comparison     472825
[x] Slevomat  Control structures  Disallow yoda comparison disallowed yoda comparison  106476
[ ] Internal                      Line endings                                         1831
[ ] Internal                      No code found                                        1762
[ ] Internal                      Exception                                            4
[ ] Internal                      Tokenizer                                            1
--------------------------------------------------------------------------------------------
A TOTAL OF 582899 SNIFF VIOLATIONS WERE FOUND IN 6 SOURCES
--------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SOURCES AUTOMATICALLY (578880 VIOLATIONS IN TOTAL)
--------------------------------------------------------------------------------------------

Time: 8 mins, 5.99 secs; Memory: 58.7MB
```

**Plugins 70k to 80k:**
```
➜  phpcs git:(master) ✗ cat 70000_80000.txt

PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
-----------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                 COUNT
-----------------------------------------------------------------------------
[x] Slevomat  Control structures  Require yoda comparison required yod  549068
[x] Slevomat  Control structures  Disallow yoda comparison disallowed   120069
[ ] Internal                      No code found                         1731
[ ] Internal                      Line endings                          1428
[ ] Internal                      Exception                             6
-----------------------------------------------------------------------------
A TOTAL OF 672302 SNIFF VIOLATIONS WERE FOUND IN 5 SOURCES
-----------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SOURCES AUTOMATICALLY (668881 VIOLATIONS IN TOTAL)
-----------------------------------------------------------------------------

Time: 9 mins, 26.69 secs; Memory: 63.78MB

```

**Plugins 80k to 82450:**
```
➜  phpcs git:(master) ✗ cat 80000_82450.txt 

PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
--------------------------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                                COUNT
--------------------------------------------------------------------------------------------
[x] Slevomat  Control structures  Require yoda comparison required yoda comparison     166025
[x] Slevomat  Control structures  Disallow yoda comparison disallowed yoda comparison  38594
[ ] Internal                      No code found                                        536
[ ] Internal                      Line endings                                         251
[ ] Internal                      Exception                                            1
--------------------------------------------------------------------------------------------
A TOTAL OF 205407 SNIFF VIOLATIONS WERE FOUND IN 5 SOURCES
--------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SOURCES AUTOMATICALLY (204464 VIOLATIONS IN TOTAL)
--------------------------------------------------------------------------------------------

Time: 3 mins, 38.64 secs; Memory: 24.75MB
```
</details>

This is the final result:

- Total Conditions: 4 610 671 (100%)
- **Normal Conditions: 3 780 281 (81.98%)**
- **Yoda Conditions: 830 390 (18.02%)**

**My thoughts:**

Looking at the data, I don't see a cascading effect on the entire world wide web that could bring doom and calamity to the internet if we stop using Yoda in Core. [The Codex itself](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#yoda-conditions) has a funny statement on the "Yoda" section: `A little bizarre, it is, to read. Get used to it, you will.`, which seems to go against what it says just a couple lines below, on the "Clever Code" section: `In general, readability is more important than cleverness or brevity.`.

To be quite frank, I think I'm dumber than most people. I really do. I don't have a very high cognitive capacity. That's why for me it's very important that the code is easy to read, follow, and understand. My personal opinion is that it compromises readability, and this is pretty much the consensus of Yoda conditions on the internet for the majority. People will not agree on everything, but we can pretty much confidently say that the majority prefers regular conditions, so I think this is a pretty pertinent discussion to have.

**Comment by JPry** [comment]
After commenting before in favor of keeping Yoda conditions, I feel like my stance has changed slightly. I think the best way forward is to:

1. Allow both styles of conditionals, Yoda and "regular". 
2. As initially described, an assignment within a conditional should be an error.
3. Loose (`==`) operators should be a warning, as they are likely to be the source of sneaky bugs.

For item 1, I think that while the discussion around readability is completely relevant, the fact remains that we're talking about comparing two values. Is it really a readability issue to distinguish between `'a' === $b` and `$b === 'a'`? Comparisons like this should be equally easy (or difficult) for a developer to parse regardless of which item is first.

To Yoda or not to Yoda feels like it is getting far more controversial than it really deserves, and perhaps the best solution is to switch from enforcing it to taking no stance at all on it. Instead, let's just focus on truly solving the problem that Yoda conditions originally intended to solve: assignments within conditionals.

**Comment by JJJ** [comment]
Thanks @Luc45 💜 I wasn't intentionally fear-mongering, but certainly unintentionally, and the data is extremely helpful to see. Relatedly, I hadn't deeply considered what the greater PHP community had decided, so that's also compelling.

Consider me persuaded. Please accept my embarrassed apologies for disrupting progress here, and you have my gratitude for being patient with my resistance.

**Comment by Luc45** [comment]
@JJJ it is with good intentions and divergent opinions that we get to the best solutions, always. If we all agreed, we would be missing the opportunity to seeing things from another angle, which could be dangerous. So thank you for disagreeing before, and for being open to the discussion.

**Comment by kraftbj** [comment]
Wanted to voice support for @JPry's comment above
> Allow both styles of conditionals, Yoda and "regular".
> As initially described, an assignment within a conditional should be an error.
> Loose (==) operators should be a warning, as they are likely to be the source of sneaky bugs.

I don't care if WPCS does end up banning yoda conditionals or not, but I think the initial step of not trying to force code to use yoda is a good step forward.

**Comment by dingo-d** [comment]
In the context of WordPress Core, allowing both is a disaster IMO. Either use one or the other. Allowing both to be in the core just adds additional confusion for new developers to the project. Should it be Yoda or not? When? Why?

We should be consistent across the codebase.

I think people confuse WPCS standards as set in stone when it comes to building their own themes or plugins. And that's not the case. You can use whatever you want in your project. You can use parts of the WPCS ruleset, remove some, leave the one you want.

Personally, I only use some functional aspects of WPCS in my projects, while removing almost all of the styling CS, because I prefer PSR-12 and PSR-4 for autoloading.

I think this issue could be talked about in a meeting and discussed in official make posts for more visibility. Maybe throw in a poll to get the wider audience reach.

**Comment by jrfnl** [comment]
> Otherwise, $a === 'b' is just as easy to understand as 'b' === $a.

As a person with dyslexia, I strongly disagree with this, the second form takes me about three times as much time to process as the first.

I do agree that the original goal of why Yoda was initially enforced has been superseded and that that should be addressed first.

As for consistency in the WP Core code base after the Yoda restriction has been removed, let's leave that for a future iteration.

For plugin/theme repos, they can choose their own rules, so even when WPCS would remove the rule, there are sniff available (in PHPCS itself/other external standards) to either enforce no Yoda or to still enforce Yoda, which a plugin/theme can elect to add to their custom ruleset.

**Comment by andronocean** [comment]
I'd like to add another reason to remove the Yoda rule: it makes the sniffs noisier for no _useful_ reason.

The sniffs only help users who install them, so anyone who doesn't and isn't contributing to Core is out of scope. If someone has installed the sniffs, they can get instant feedback on errors. The most useful feedback is warnings about assignment-in-condition and loose comparisons: those are the specific things that cause problems, and that info is helpful to novice and experienced  devs alike.

Enforcing Yoda doesn't add anything to this — it just generates more warnings, more noise for the developer to process, and hence more overwhelm and a longer learning curve for new devs.

**Comment by andronocean** [comment]
> I think you're confusing code style checks with static code analysis, such as PHPStan.

How so? This is all checkable through PHPCS. I can get alerts from it as I code with something like the [PHP Sniffer & Beautifier](https://marketplace.visualstudio.com/items?itemName=ValeryanM.vscode-phpsab&ssr=false#overview) extension in VS Code. No static analysis necessary, afaik.

**Comment by Luc45** [comment]
> Just a FYI: PHPCS === static analysis, just based on a different entry point to the code compared to PHPStan.

True that. My reasoning when I said that was because of the intent... PHPCS uses static analysis to check code style, while PHPStan uses static analysis to check logic errors, so if by "useful error" he means "flagging logic errors in my code", PHPCS is not the tool for that, as PHPCS only enforces code style conventions that shouldn't affect the logic of the code, no matter how the code style looks like.

`<offtopic/>`

**Comment by Luc45** [comment]
As for next steps, what about?
- Instead of enforcing Yoda, ban Yoda.
- Increase the code style version to `^2.4.0`, so projects that are already written in Yoda can lock to `^2.3.0` to avoid having to refactor to comply with the code style checks and keep CI happy.
- Add documentation on how to convert a codebase from Yoda to non-yoda, which is simply: Run PHPCBF with PHP-CS-Fixer `yoda_style` rules configured to enforce non-yoda [[Link](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/doc/rules/control_structure/yoda_style.rst#example-4)]

**Comment by jrfnl** [comment]
> > Just a FYI: PHPCS === static analysis, just based on a different entry point to the code compared to PHPStan.
> 
> True that. My reasoning when I said that was because of the intent... PHPCS uses static analysis to check code style, while PHPStan uses static analysis to check logic errors, so if by "useful error" he means "flagging logic errors in my code", PHPCS is not the tool for that, as PHPCS generally speaking is meant to enforce code style conventions that shouldn't affect the logic of the code, no matter how the code style looks like.
> 
> `<offtopic/>`

While maybe off topic, it does a disservice to PHPCS, which can do a lot more than just check for code style convention. Think: the [PHPCompatibility](https://github.com/PHPCompatibility/PHPCompatibility) standard and sniffs in the CodeAnalysis category.

Of course, there are limitations, but don't underestimate PHPCS.


As for next steps, see my previous post about this and the original topic.

Most likely, we'll need a Make post before we can put this into effect anyhow.

Other than that:
* The next version of WPCS will be 3.0.0, there is not going to be a 2.4.0.
* Regarding fixing - you are confusing two different tools - CS-Fixer is completely unrelated to PHPCS/PHPCBF and these tools don't work with the rulesets from each other/cannot be used interchangably.

**Comment by Luc45** [comment]
@jrfnl good points about the PHPCompatibility capabilities of PHPCS. A while back I needed to convert a project from Yoda to non-yoda, and PHP-CS-Fixer was the only tool that I found that could do it ([Link](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5293)). Indeed, I remember now that it was a separate tool from PHPCBF.

**Comment by Luc45** [comment]
> @Luc45 As described in the original topic, there is a PHPCS sniff in the Slevomat standard which can do this perfectly fine as well.

You're right, thanks. I'm replying here while working on other stuff and am a little bit distracted. I remember now that I have used that Slevomat standard when counting Yoda vs Non-Yoda in the WordPress Plugin ecosystem, actually... I agree that bumping to ^3.0.0 makes sense. Do you agree about banning assignments in conditions?

**Comment by Luc45** [comment]
A good example of where Yoda can get really confusing: Inverted comparison operators:

![image](https://user-images.githubusercontent.com/9341686/169100225-b2a84c92-8c63-4c1b-bb16-3e18a8e315d8.png)

Friends, what should we do next to bring this proposal to the table? Does anyone oppose getting rid of Yoda?

**Comment by 2ndkauboy** [comment]
> A good example of where Yoda can get really confusing: Inverted comparison operators:

These were never included in the Yoda condition rules.

**Comment by Luc45** [comment]
> These were never included in the Yoda condition rules.

Ah, you're right, thanks for the clarification. I have verified this in multiple ways and that statement does hold true.

<details>
  <summary>Yoda and Numeric Comparison verification</summary>
As expected, Yoda rules does not enforce any kind of inverted order for numeric comparisons, in multiple scenarios as tested below. None of these scenarios are flagged by Yoda rules:

`/vendor/bin/phpcs foo.php --standard=WordPress --sniffs=WordPress.PHP.YodaConditions`

```php
<?php

$count = 21;
$target = 20;
$getCount = static function () {
	return 21;
};
$getTarget = static function () {
	return 20;
};

// Numeric comparison with int - Inverted
if ( 20 <= $count ) {
	// no-op
}
// Numeric comparison with int
if ( $count > 20 ) {
	// no-op
}

// Numeric comparison with variable - Inverted
if ( $target <= $count ) {
	// no-op
}
// Numeric comparison with variable
if ( $count > $target ) {
	// no-op
}

// Numeric comparison with callable - Inverted
if ( $getTarget <= $count ) {
	// no-op
}
// Numeric comparison with callable
if ( $count > $getTarget ) {
	// no-op
}

// Numeric comparison with callable (variation) - Inverted
if ( $target <= $getCount ) {
	// no-op
}
// Numeric comparison with callable (variation)
if ( $getCount > $target ) {
	// no-op
}

// Numeric comparison with callable (variation 2) - Inverted
if ( $getTarget <= $getCount ) {
	// no-op
}
// Numeric comparison with callable (variation 2)
if ( $getCount > $getTarget ) {
	// no-op
}
```
</details>

Even though the enforcement is not there at a technical level, this popped up on a code review for me on a codebase that has Yoda enforced, I have asked the developer if he had struggled to write that line as much as I did to read it and he vented with me that he had struggled a lot. I will inform him that he does not need to invert numeric comparisons.

Now, the most important: **Does anyone oppose @jrfnl proposal?** I would like to pledge that we refrain all further replies to this lengthy thread to the resolution of this issue, so that we can get to a conclusion.

**Comment by jrfnl** [comment]
As this has been a long-standing issue with widely opposing views and lots of people with firm opinions involved, I believe a post on Make to announce our intention to change this is needed to move this forward.

Volunteers to write the post very welcome. I'm happy to review it before publication.

In my opinion, these are the action steps needed to get this sorted (in order):
- [ ] Publish announcement of the intention to change this rule on Make Core, including what the "new" rules will be.
- [ ] If the majority of responses to the Make post is positive, send in a [PR to the Core CS handbook repo](https://github.com/WordPress/wpcs-docs) to update the official standards.
- [ ] After the handbook has been changed, send in a PR to this repo to make the changes as originally proposed.

**Comment by 2ndkauboy** [comment]
@Luc45 I don't oppose with the initial idea of @jrfnl to not enforce the Yoda condition any more (although I personally prefer it). I would just ask us not to "ban", disallow it, as some other people recommended in the comments. Basically what @jrfnl mentioned in this comment: https://github.com/WordPress/WordPress-Coding-Standards/issues/1624#issuecomment-470821000

**Comment by Luc45** [comment]
> @Luc45 I don't oppose with the initial idea of @jrfnl to not enforce the Yoda condition any more (although I personally prefer it). I would just ask us not to "ban", disallow it, as some other people recommended in the comments. Basically what @jrfnl mentioned in this comment: https://github.com/WordPress/WordPress-Coding-Standards/issues/1624#issuecomment-470821000

I understand, and this is indeed a very delicate topic. As banning Yoda might seem as unfair to some as enforcing  is to others.

Considering, and feeling, and trying to be fair with this, I am considering an equation like this: 

```
$ecosystem = $developer_happiness + $codebase_consistency;
```

By enforcing Yoda, we can, with the data we have, (count of yoda vs regular and count of reactions in this proposal) infer that it decreases `$developer_happiness`. 

```
// Considering:
// Yoda: 20% of conditionals found in large-scale sample
// Regular: 80% of conditionals found in large-scale sample

$ecosystem = 0.2 + 1; // 1.2 Enforcing Yoda
$ecosystem = 0.8 + 1; // 1.8 Banning Yoda
$ecosystem = 1 + ?; // ? Allowing Both
```

The critical point is that if we don't ban Yoda, we will never have enough incentive to get rid of it, thus decreasing `$codebase_consistency` dramatically over time, leading to a lower score of `$ecosystem`.

If you are passionate about Yoda, you can always disable the PHPCS sniff or even enforce Yoda in your project.

With that said, I think we can save our most brilliant discussions to the Make post, which I think is the most efficient path moving forward.

---

> Volunteers to write the post very welcome.

Your proposal in this thread is already a very nice starting point to a post. I am sorry for not being able to help on that, as I'm going through one of those "crunch times" that happens every time you have a deadline and lots of technical work to do, which is a feeling I think most software developers know very well.

Someone...? :disguised_face:

**Comment by Luc45** [comment]
> @Luc45 I don't oppose with the initial idea of @jrfnl to not enforce the Yoda condition any more (although I personally prefer it). I would just ask us not to "ban", disallow it, as some other people recommended in the comments. Basically what @jrfnl mentioned in this comment: https://github.com/WordPress/WordPress-Coding-Standards/issues/1624#issuecomment-470821000

I think it's fine for us to have different opinions. As a collective we are always making choices, and divergence is a natural part of this process. Our sum of choices over time defines the success or failure of the collective enterprise, so it's a big responsibility to have opinions and defend them, which is only worthy as long as we still believe in what we are saying.

With that in mind, and aware that this divergence exist, I find myself in a crossroad: We can either write the Make post and bring further discussions there, as a definitive, formal place. Or only write the Make post once we have reached a consensus on whether to ban yoda or be indifferent about it.

I think the concept of a "Consensus" in a collective can actually be harmful to strive for. As to achieve consensus on every decision, we might often need to make suboptimal concessions for immediate happiness that are not healthy to the collective as a whole, like politics.

With that said, and taking upon the responsibility of keeping this thread on track so that we can keep a level of efficiency towards the best outcome for the issue at hand, I ask:

@2ndkauboy (and others), why should we allow Yoda in WordPress Coding Standards ^3? Don't you think it would decrease the codebase consistency, which is one of the main goals of a code style convention?

**Comment by 2ndkauboy** [comment]
Why we should allow it? That's easy to answer: because we have enforced it in the past. Now banning it would make the vote more inconsistent, because we would need to change it on all occurrence, this would be a huge change. And we also have to get it out of the muscle memory of every WordPress developer, which might be just as hard 😁

**Comment by Luc45** [comment]
For the sake of efficiency, let's focus on making this the final discussion before the Make post.
A) **To ban Yoda.**
B) **To be indifferent about Yoda.**

I prefer A, but I understand your argument in regards to "time to adapt".

With that said, I still think A is the best path moving forward, given a few criteria are met:

**1: Codebases that need more time can always stick to WPCS ^2 for a while.**

**2: Simple instructions on how to automatically swap Yoda for regular conditions on existing codebases are documented.**

Swapping Yoda for Regular on a codebase should ideally be a one-liner, something like:

`./vendor/bin/phpcbf src --standard=WordPress --sniffs=WordPress.PHP.YodaConditions`

_(Considering `Slevomat Coding Standard` is added as a dependency of WPCS ^3 as proposed in the original thread)_

In a large codebase with many open PRs branched off before the coding standard has been updated, this is one approach to avoid merge conflicts with open PRs:

1. Create a branch with only the version bump:
- `git checkout trunk`
- `git branch -b wpcs-3`
- `composer require wp-coding-standards/wpcs ^3 --dev`
- `git add . && git commit -m "WPCS ^3" && git push`

2. Create another branch for trunk fixed by PHPCBF:
- `git checkout trunk`
- `git branch -b wpcs-phpcbf-3`
- `git merge wpcs-3`
- `./vendor/bin/phpcbf src --standard=WordPress --sniffs=WordPress.PHP.YodaConditions`
- `git add . && git commit -m "PHPCBF WPCS ^3 && git push`

3. Foreach PR:
- `git checkout my-pr`
- `git merge wpcs-3`
- `./vendor/bin/phpcbf src --standard=WordPress --sniffs=WordPress.PHP.YodaConditions`
- `git add . && git commit -m "PHPCBF WPCS ^3" && git push`

Now trunk and all open PRs have updated their codebases to regular conditions without any merge conflicts or manual labor.

With that said, this is what I personally defend to keep the codebase consistent between new and existing code. These are my arguments, and if we agree to disagree, that's fine. I am proud that we are able to have a rational discussion about Coding Standads at all, instead of going against each other with pitchforks and torches. This is the sign of a strong community, where everyone is comfortable coming forward and defend their opinion without being torn apart for it.

I'm looking forward for the conclusion and a hero that will write that Make post, be it as it may.

**Comment by dingo-d** [comment]
@2ndkauboy 

> It would also break many pending patches to core with a Yoda condition in place.

Not it wouldn't break anything if it was removed. No sniff is checking the absence of Yoda, only its enforcement. So code written with Yoda style or without it, if you don't include Yoda won't report: `Missing Yoda condition this statement is` 😂 

Also, the argument that a lot of plugins/themes use Yoda or not, is not a concern for *WordPress Core*. A lot of plugins follow PSR-12 (or 2), but that doesn't mean we'd need to enforce curly braces on the separate line in the WordPress Core.

**Comment by 2ndkauboy** [comment]
@dingo-d I'm talking about any patch on Trac/GitHub waiting to be merged with changes including a Yoda condition.

As far as we care about the WordPress Core: it uses only Yoda conditions. Plugins can and will use whatever syntax they want to use.

**Comment by 2ndkauboy** [comment]
@jrfnl thanks for pointing me to that comment. I haven't recognized it really.

I did a quick check on a fresh copy of WordPress 5.9.3 if I exclude all external dependencies/libraries, I get the following numbers:

```
PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
--------------------------------------------------------------------------------------------
    STANDARD  CATEGORY            SNIFF                                                COUNT
--------------------------------------------------------------------------------------------
[x] Slevomat  Control structures  Disallow yoda comparison disallowed yoda comparison  4621
[x] Slevomat  Control structures  Require yoda comparison required yoda comparison     302
[ ] WordPres  PHP                 Yoda conditions not yoda                             9
--------------------------------------------------------------------------------------------
```

Those 9 `WordPress.PHP.YodaConditions.NotYoda` violations come from the three "(ms-)deprecated.php" files. But what I also realized is that the `SlevomatCodingStandard.ControlStructures.RequireYodaComparison` would also flag a comparsion of a function to a string (like `strpos( $permalink_structure, '%year%' ) !== false`) but `WordPress.PHP.YodaConditions.NotYoda` does not flag that as a violation.

The summary also shows the number of violations, not the number of plugins, so we can't really tell from that summary how many plugins do use Yoda conditions and how many don't.

If we care about WordPress Core only, the Core is exclusively using Yoda conditions (except for those 9 violations in deprecated functions). If the number of violations of `SlevomatCodingStandard.ControlStructures.DisallowYodaComparison` is correct though, this would mean that Core would need a huge commit of up to 4621 changed lines.

So I would strongly suggest to go with your initial approach of just removing the sniff and going with **B) To be indifferent about Yoda** as @Luc45 mentioned.

The bigger issues are `WordPress.CodeAnalysis.AssignmentInCondition`, `WordPress.PHP.StrictComparisons` and `WordPress.PHP.StrictInArray`, but not whether we want to ban the Yoda condition from WordPress Core.

**Comment by dingo-d** [comment]
Again, I wouldn't include Slevomats Yoda check at all in core. So no patches would fail. No notices in the core would be thrown.

>  So I would strongly suggest to go with your initial approach of just removing the sniff and going with B) To be indifferent about Yoda as @Luc45 mentioned, or even keeping the sniff and move the other ones to the Core ruleset.

This is my stance as well. Remove the sniff, don't enforce it, but don't check if Yoda code exists or not in the codebase. If it exists, ok, if the new patches are not using it, also ok.

**Comment by dingo-d** [comment]
I've written a proposal on the core blog, it's still in draft mode.

@jrfnl @GaryJones I'd love to get your feedback [on it](https://make.wordpress.org/core/?p=96537&preview=true) when you'll find the time.

**Comment by dingo-d** [comment]
The article is live: https://make.wordpress.org/core/2022/06/14/upcoming-disallow-assignments-in-conditions-and-remove-the-yoda-condition-requirement-for-php/

**Comment by jrfnl** [comment]
Please participate in this unscientific poll on Yoda conditions: https://twitter.com/jrf_nl/status/1537015003546542080

**Comment by dingo-d** [comment]
The loose comparisons and assignments in conditions checks are in the core ruleset, but despite the overwhelming argument for removing Yoda conditions by the community, it was decided to leave those in the core ruleset.

If in the future we choose to revisit the Yoda conditions, we'll open a new issue.
